### PR TITLE
AWS SQS Job Locker to Prevent Duplication

### PR DESCRIPTION
**Pull request by fustundag**
While using AWS SQS Standart Queue, sometimes same job sent more than once. To prevent this, added locking logic to SqsQueue getNext method.

http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/standard-queues.html#standard-queues-at-least-once-delivery

### CODE REVIEW

**Comment by fustundag** [comment]
Any comment on pull request? We need this pull request at our projects

File: src/Simpleue/Locker/LockerAbstract.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace Simpleue\Locker;
+
+abstract class LockerAbstract implements LockerInterface
```
Comment (javibravo) [inline src/Simpleue/Locker/LockerAbstract.php]: I would suggest removing "Abstract" suffix for the class, there is no need to use it in the class name.

File: src/Simpleue/Locker/LockerInterface.php [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Simpleue\Locker;
+
+interface LockerInterface
```
Comment (javibravo) [inline src/Simpleue/Locker/LockerInterface.php]: Suffix "Interface" should be removed. It is not used in other interfaces as "Job" or "Queue".

File: src/Simpleue/Locker/MemcachedLocker.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        if (!$job) {
+            throw new \RuntimeException('Job for lock is invalid!');
+        }
+        $status = $this->memcached->add(
+            $this->getJobUniqId($job),
+            time() + $timeout + 1,
+            $timeout
+        );
+        if ($status) {
```
Comment (javibravo) [inline src/Simpleue/Locker/MemcachedLocker.php]: "add" method from Memcached returns a boolean value, why not return it directly?

```
return $status
```
or 
```
return $this->memcached->add(
    $this->getJobUniqId($job),
    time() + $timeout + 1,
    $timeout
);
```

File: src/Simpleue/Locker/RedisLocker.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            $key,
+            time() + $timeout + 1,
+            array('nx', 'ex' => $timeout)
+        );
+        if ($status) {
+            return true;
+        }
+
+        $currentLockTimestamp = $this->redis->get($key);
+        if ($currentLockTimestamp>time()) {
```
Comment (javibravo) [inline src/Simpleue/Locker/RedisLocker.php]: Missing spaces before and after ">".

File: src/Simpleue/Locker/RedisLocker.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            return false;
+        }
+        $oldLockTimestamp = $this->redis->getSet($key, (time() + $timeout + 1));
+        if ($oldLockTimestamp>time()) {
```
Comment (javibravo) [inline src/Simpleue/Locker/RedisLocker.php]: Missing spaces before and after ">".

File: src/Simpleue/Worker/QueueWorker.php [hunk 6]
```diff
@@ -71,7 +71,6 @@ public function start()
                 $job = $this->queueHandler->getNext();
             } catch (\Exception $exception) {
                 $this->log('error', 'Error getting data. Message: '.$exception->getMessage());
-                $this->queueHandler->error(false, $exception);
```
Comment (javibravo) [inline src/Simpleue/Worker/QueueWorker.php]: Good catch.

File: src/Simpleue/Worker/QueueWorker.php [hunk 7]
```diff
@@ -157,7 +156,7 @@ private function manageJob($job)
             }
         } catch (\Exception $exception) {
             $this->log('error', 'Error Managing data. Data :'.$this->queueHandler->toString($job).'. Message: '.$exception->getMessage());
-            $this->queueHandler->error($job, $exception);
```
Comment (javibravo) [inline src/Simpleue/Worker/QueueWorker.php]: Good catch.

**Comment by fustundag** [comment]
@javibravo I fixed with your comments.
LockerInterface changed to Locker, LockerAbstract changed to BaseLocker because of name conflict.

**Comment by javibravo** [comment]
Released on [2.1.0](https://github.com/javibravo/simpleue/releases/tag/2.1.0)

**Pull request merged**